### PR TITLE
[Feature] Bits Bytes IR core function

### PR DESCRIPTION
**Pull request by gluax**
## Motivation

To bits and to bytes core function calls in snarkvm for leo.

Right now on the master branch as far as I could tell only implements to_bits and from_bits for integer types. So the core calls exist on the other types to error out for now till those are in. 

## Test Plan

The IR is already thoroughly tested and all tests currently pass.

## Related PRs

Should be merged into #416, before 416 is merged into master, or merged into master after 416 is merged into master.

### CODE REVIEW

**Review by howardwu — DISMISSED** [review]
Before proceeding further on bits and bytes, we need to change convention to `to_bits_le` and `to_bytes_le`, or `_be` if that's what's going on here

File: eval/src/value/value.rs [hunk 1]
```diff
@@ -106,6 +109,36 @@ impl<F: PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
             (_, _) => false,
         }
     }
+
+    pub fn to_bits_le(&self) -> Result<Vec<Boolean>> {
+        use ConstrainedValue::*;
+
+        match self {
+            Address(_) => Err(anyhow!("the type `address` does not implement the to_bytes method")),
```
Comment (collinc97) [inline eval/src/value/value.rs:117]: should these error messages read "... to_bits_le method"

File: eval/src/value/value.rs [hunk 2]
```diff
@@ -106,6 +109,36 @@ impl<F: PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
...
+            Array(_) => Err(anyhow!("the type `array` does not implement the to_bytes method")),
+            Tuple(_) => Err(anyhow!("the type `tupple` does not implement the to_bytes method")),
+        }
+    }
+
+    pub fn to_bytes_le(&self) -> Result<Vec<UInt8>> {
+        use ConstrainedValue::*;
+
+        match self {
+            Address(_) => Err(anyhow!("the type `address` does not implement the to_bytes method")),
```
Comment (collinc97) [inline eval/src/value/value.rs:132]: similarly "... to_bytes_le method"

**Review by collinc97 — CHANGES_REQUESTED** [review]

File: eval/src/value/value.rs [hunk 3]
```diff
@@ -106,6 +109,36 @@ impl<F: PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
...
+
+        match self {
+            Address(_) => Err(anyhow!("the type `address` does not implement the to_bits_le method")),
+            Boolean(_) => Err(anyhow!("the type `bool` does not implement the to_bits_le method")),
+            Char(_) => Err(anyhow!("the type `char` does not implement the to_bits_le method")),
+            Field(_) => Err(anyhow!("the type `field` does not implement the to_bits_le method")),
+            Group(_) => Err(anyhow!("the type `group` does not implement the to_bits_le method")),
+            Integer(integer) => Ok(integer.to_bits_le()),
+            Array(_) => Err(anyhow!("the type `array` does not implement the to_bits_le method")),
+            Tuple(_) => Err(anyhow!("the type `tupple` does not implement the to_bits_le method")),
```
Comment (collinc97) [inline eval/src/value/value.rs:124]:
```suggestion
            Tuple(_) => Err(anyhow!("the type `tuple` does not implement the to_bits_le method")),
```

File: eval/src/value/value.rs [hunk 4]
```diff
@@ -106,6 +109,36 @@ impl<F: PrimeField, G: GroupType<F>> ConstrainedValue<F, G> {
...
+
+        match self {
+            Address(_) => Err(anyhow!("the type `address` does not implement the to_bytes_le method")),
+            Boolean(_) => Err(anyhow!("the type `bool` does not implement the to_bytes_le method")),
+            Char(_) => Err(anyhow!("the type `char` does not implement the to_bytes_le method")),
+            Field(_) => Err(anyhow!("the type `field` does not implement the to_bytes_le method")),
+            Group(_) => Err(anyhow!("the type `group` does not implement the to_bytes_le method")),
+            Integer(_) => Err(anyhow!("the type `int` does not implement the to_bytes_le method")),
+            Array(_) => Err(anyhow!("the type `array` does not implement the to_bytes_le method")),
+            Tuple(_) => Err(anyhow!("the type `tupple` does not implement the to_bytes_le method")),
```
Comment (collinc97) [inline eval/src/value/value.rs:139]:
```suggestion
            Tuple(_) => Err(anyhow!("the type `tuple` does not implement the to_bytes_le method")),
```

**Review by collinc97 — DISMISSED** [review]
grammar changes

**Comment by collinc97** [comment]
@gluax can we procedurally generate all these methods instead of hard coding them all?
We're expecting ourselves to go back and implement all of these methods at a later date under current assumptions.
But I fear those assumptions may change once we start designing the new programming model for testnet2.

**Comment by gluax** [comment]
> 
> 
> @gluax can we procedurally generate all these methods instead of hard coding them all? We're expecting ourselves to go back and implement all of these methods at a later date under current assumptions. But I fear those assumptions may change once we start designing the new programming model for testnet2.

We cannot as the function does not have the context to what type it should be outputting at the IR for a generic solution. So we need a method per on the from ones.

**Review by 0rphon — APPROVED** [review]

**Comment by Protryon** [comment]
Can we use some rust macros to generate the boilerplate rather than making the IR dynamically typed? I.e. like we do for a `Integer` impls in `snarkvm_algorithms`.

**Pull request closed without merging**